drivers/serial/uart_mcux_flexcomm.c

```diff
@@ -342,6 +379,636 @@ static int mcux_flexcomm_uart_config_get(const struct device *dev,
 }
 #endif /* CONFIG_UART_USE_RUNTIME_CONFIGURE */
 
+#ifdef CONFIG_UART_ASYNC_API
+/* This function is called by this driver to notify user callback of events */
+static void async_user_callback(const struct device *dev,
+					      struct uart_event *evt)
+{
+	const struct mcux_flexcomm_data *data = dev->data;
+
+	if (data->async_callback) {
+		data->async_callback(dev, evt, data->async_cb_data);
+	}
+}
+
+static int mcux_flexcomm_uart_callback_set(const struct device *dev,
+					   uart_callback_t callback,
+					   void *user_data)
+{
+	struct mcux_flexcomm_data *data = dev->data;
+
+	data->async_callback = callback;
+	data->async_cb_data = user_data;
+
+	return 0;
+}
+
+static int mcux_flexcomm_uart_tx(const struct device *dev, const uint8_t *buf,
+				 size_t len, int32_t timeout)
+{
+	const struct mcux_flexcomm_config *config = dev->config;
+	struct mcux_flexcomm_data *data = dev->data;
+	int ret = 0;
+
+	if (config->tx_dma.dev == NULL) {
+		return -ENODEV;
+	}
+
+	unsigned int key = irq_lock();
+
+	/* Getting DMA status to tell if channel is busy or not set up */
+	struct dma_status status;
+
+	ret = dma_get_status(config->tx_dma.dev, config->tx_dma.channel, &status);
+
+	if (ret < 0) {
+#ifdef CONFIG_DMA_MCUX_LPC
+		if (ret == -EACCES) {
+			/* In this case this driver knows which channel is valid,
+			 * so a return of -EACCES can only mean some virtual channel
+			 * in the LPC DMA driver is not set up, so just clarify it's okay
+			 * to proceed to set up the UART RX
+			 */
+			status.busy = false;
+		} else {
+			return ret;
+		}
+#else
+		return ret;
+#endif /* CONFIG_DMA_MCUX_LPC */
+	}
+
+	/* There is an ongoing transfer */
+	if (status.busy) {
+		return -EBUSY;
+	}
+
+	/* Disable TX DMA requests for uart while setting up */
+	USART_EnableTxDMA(config->base, false);
+
+	/* Set up the dma channel/transfer */
+	data->tx_data.xfer_buf = buf;
+	data->tx_data.xfer_len = len;
+	data->tx_data.active_block.source_address = (uint32_t)buf;
+	data->tx_data.active_block.dest_address = (uint32_t) &config->base->FIFOWR;
+	data->tx_data.active_block.block_size = len;
+	data->tx_data.active_block.next_block = NULL;
+
+	ret = dma_config(config->tx_dma.dev, config->tx_dma.channel,
+				(struct dma_config *) &config->tx_dma.cfg);
+	if (ret) {
+		return ret;
+	}
+
+	/* Enable interrupt for when TX fifo is empty (all data transmitted) */
+	config->base->FIFOINTENSET |= USART_FIFOINTENSET_TXLVL_MASK;
```

Comment:
Two comments here:
1. Could we use the TX IDLE interrupt to signal all data is transmitted? It is not clear to me from the hardware RM if the TXLVL interrupt would fire before the transfer is complete when more than 16 bytes are sent (as in this case the TX FIFO would fill, drain, and then the DMA would fill it again). As a side benefit, this way we can configure the TXLVL watermark higher than 0, if we want to set up the USART this way in the future.
2. Why not use the HAL API here?

Reply:
how would TX idle be different from  TXLVL, does the transmitter not idle when the fifo is empty and the uart has no data to send?

Reply:
The reference manual for the USART made it unclear to me how TXLVL would function when using DMA. In the case of DMA, it seems that when the TX FIFO is drained the TXLVL interrupt will be asserted (or the DMA request, if the interrupt signal is disabled. My concern was that the TX FIFO might become drained in some cases prior to the DMA writing new data, resulting in multiple TX done events being sent (one per interrupt) however, I have tested for this issue using PR locally, and that does not seem to occur. So I'm going to go ahead and resolve this comment.